server/src/test/java/org/elasticsearch/cluster/metadata/MetadataIndexStateServiceBatchingTests.java

```diff
@@ -139,7 +144,10 @@ public void testBatchCloseIndices() throws Exception {
             assertThat(indexMetadata.getState(), is(State.CLOSE));
         }
 
-        clusterService.removeListener(assertingListener);
+        clusterService.removeListener(closedIndicesStateListener);
+        observedClosedIndices.forEach(
+            indices -> assertThat("unexpected closed indices: " + Arrays.toString(indices), indices.length, oneOf(0, 3))
+        );
```

Comment:
Should we change the `closedIndexCountListener` instead?

Please correct me if I am wrong but 
```
    private static ClusterStateListener closedIndexCountListener(int closedIndices) {
        return event -> assertThat(event.state().metadata().getConcreteAllClosedIndices(), oneOf(hasSize(0), hasSize(closedIndices)));
    }
```

should list the contents of the collection if that failed the assertion?
This way it will be applicable for all testcases using that.

Reply:
The issue here is _not_ that we don't get enough information from this assertion. Based on the failure [logs](https://gradle-enterprise.elastic.co/s/vn2vj6dmamixc/tests/task/:server:test/details/org.elasticsearch.cluster.metadata.MetadataIndexStateServiceBatchingTests/testBatchCloseIndices?focused-execution=1&top-execution=1#L10-L12), it was actually pretty clear that the actual indices were just `test-1`. The problem is that throwing assertion error here stops the cluster state processing which in turn stops the underlying exception to be surfaced via `future.get()`. I am fairly certain the underlying exception is thrown inside the [closeRoutingTable](https://github.com/elastic/elasticsearch/blob/7ef52ec3a34190a1dd38e40b9e71a541105c42fb/server/src/main/java/org/elasticsearch/cluster/metadata/MetadataIndexStateService.java#L809) method. So the main purpose of this PR is to ensure it is logged when the test fails again. I hope this makes sense.